File: TMessagesProj/src/main/java/ua/itaysonlab/tgkit/preference/types/TGKitListPreference.java
package ua.itaysonlab.tgkit.preference.types;

import android.app.Activity;
import android.view.View;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;
import androidx.core.util.Pair;

import java.util.ArrayList;
import java.util.List;

import bruhcollective.itaysonlab.airui.miui.popup.DropdownPopup;
import bruhcollective.itaysonlab.airui.miui.popup.adapter.SelectiveMenuAdapter;
import bruhcollective.itaysonlab.airui.miui.popup.adapter.content.DoubleLineContentAdapter;
import kotlin.Triple;
import kotlin.Unit;
import ua.itaysonlab.redesign.BaseActionedSwipeFragment;
import ua.itaysonlab.redesign.sheet.TgxMessageMenuSheetFragment;
import ua.itaysonlab.redesign.slides.TgxMessageMenuFragment;
import ua.itaysonlab.tgkit.preference.TGKitPreference;

public class TGKitListPreference extends TGKitPreference {
    public boolean divider = false;
    public TGTLContract contract;

    @Override
    public TGPType getType() {
        return TGPType.LIST;
    }

    public void callActionHueta(Activity pr, View view, float x, float y, TempInterface ti) {
        ArrayList<SelectiveMenuAdapter.Item> actions = new ArrayList<>();

        if (contract.hasIcons()) {
            for (Triple<Integer, String, Integer> act : contract.getOptionsIcons()) {
                actions.add(new SelectiveMenuAdapter.Item(String.valueOf(act.getFirst()), act.getSecond(), null, ContextCompat.getDrawable(pr, act.getThird()), contract.getValue().equals(act.getSecond())));
            }
        } else {
            for (Pair<Integer, String> act : contract.getOptions()) {
                actions.add(new SelectiveMenuAdapter.Item(String.valueOf(act.first), act.second, null, null, contract.getValue().equals(act.second)));
            }
        }

        DropdownPopup drp =  new DropdownPopup(pr, view, (position, item) -> {
            contract.setValue(Integer.parseInt(item.getId()));
            ti.update();
        });

        drp.setAdapter(new SelectiveMenuAdapter(pr, actions, new DoubleLineContentAdapter(pr, actions)));
        drp.show(View.TEXT_DIRECTION_INHERIT, View.TEXT_ALIGNMENT_TEXT_START, x, y);
    }

    public interface TGTLContract {
        void setValue(int id);
        String getValue();
        List<Pair<Integer, String>> getOptions();
        List<Triple<Integer, String, Integer>> getOptionsIcons();

        boolean hasIcons();
    }

    public interface TempInterface {
        void update();
    }
}

File: TMessagesProj/src/main/java/ua/itaysonlab/tgkit/preference/types/TGKitListPreference.kt
package ua.itaysonlab.tgkit.preference.types

import android.app.Activity
import android.content.DialogInterface
import android.view.View
import androidx.core.content.ContextCompat
import androidx.core.util.Pair
import bruhcollective.itaysonlab.airui.miui.popup.DropdownPopup
import bruhcollective.itaysonlab.airui.miui.popup.adapter.SelectiveMenuAdapter
import bruhcollective.itaysonlab.airui.miui.popup.adapter.content.DoubleLineContentAdapter
import org.telegram.ui.ActionBar.BaseFragment
import org.telegram.ui.Components.AlertsCreator
import ua.itaysonlab.catogram.CatogramConfig
import ua.itaysonlab.tgkit.preference.TGKitPreference

class TGKitListPreference : TGKitPreference() {
    var divider = false
    var contract: TGTLContract? = null

    override fun getType(): TGPType {
        return TGPType.LIST
    }

    fun callActionHueta(bf: BaseFragment, pr: Activity, view: View, x: Float, y: Float, ti: TempInterface) {
        if (CatogramConfig.useAirUiPopup) {
            val actions = ArrayList<SelectiveMenuAdapter.Item>()
            if (contract!!.hasIcons()) {
                for ((first, second, third) in contract!!.getOptionsIcons()) {
                    actions.add(SelectiveMenuAdapter.Item(first.toString(), second, null, ContextCompat.getDrawable(pr, third!!), contract!!.getValue() == second, false))
                }
            } else {
                for (act in contract!!.getOptions()) {
                    actions.add(SelectiveMenuAdapter.Item(act.first.toString(), act.second, null, null, contract!!.getValue() == act.second, false))
                }
            }

            val drp = DropdownPopup(pr, view) { position: Int, id ->
                contract!!.setValue(id.id.toInt())
                ti.update()
            }

            drp.setAdapter(SelectiveMenuAdapter(pr, actions, DoubleLineContentAdapter(pr, actions)))
            drp.show(View.TEXT_DIRECTION_INHERIT, View.TEXT_ALIGNMENT_TEXT_START, x, y)
        } else {
            var selected: Int = 0
            val titleArray = mutableListOf<String>()
            val idArray = mutableListOf<Int>()

            if (contract!!.hasIcons()) {
                contract!!.getOptionsIcons().forEachIndexed { index, triple ->
                    titleArray.add(triple.second)
                    idArray.add(triple.first)

                    if (contract!!.getValue() == triple.second) selected = index
                }
            } else {
                contract!!.getOptions().forEachIndexed { index, pair ->
                    titleArray.add(pair.second)
                    idArray.add(pair.first)

                    if (contract!!.getValue() == pair.second) selected = index
                }
            }

            val d = AlertsCreator.createSingleChoiceDialog(pr, titleArray.toTypedArray(), title, selected) { di, sel ->
                contract!!.setValue(idArray[sel])
                ti.update()
            }

            bf.visibleDialog = d

            d.show()
        }
    }

    interface TGTLContract {
        fun setValue(id: Int)
        fun getValue(): String
        fun getOptions(): List<Pair<Int, String>>
        fun getOptionsIcons(): List<Triple<Int, String, Int?>>
        fun hasIcons(): Boolean
    }

    interface TempInterface {
        fun update()
    }
}